### ISSUE TITLE
Can the internal _value of BehaviourSubject be nulled out on complete?

### ISSUE THREAD

**Issue by Wilt** [ReactiveX/rxjs#5405]
## Feature Request

Should `BehaviourSubject::_value` be nulled out on complete? Or is there a reason that it remains even after completion?
This might empty out some used memory and would allow the unused value to be garbage collected?

Maybe something in the line of:

````
  complete(): void {
    this._value = null;
    super.complete();
  }
````

At first I thought it is not nulled out because of the public `getValue()` method, but then it was clear that even this method will not return the value after completion, rendering the value useless even in this case.

Maybe I missed some other case then I would gladly hear of the reason to understand why value should remain even after calling complete.

**Comment by cartant** [comment]
No. It's not completion that closes a subject; it's calling `unsubscribe` on the subject itself:

https://github.com/ReactiveX/rxjs/blob/master/src/internal/Subject.ts#L105-L109

It would make some sense to null out the value when that's done, but, TBH, calling `unsubscribe` on subjects is pretty unusual: https://ncjamieson.com/closed-subjects/

**Comment by Wilt** [comment]
Thank you for your response, a link to a great blog post and enlightening me. <br>I am no expert on RxJs (yet :) ) and I should have looked better in the parent class before posting. I clearly misunderstood the internals of the Subject class.

It also makes me realize that completing subjects for cleaning up is not enough, I should actually even call unsubscribe on those and then I will get _"load and angry errors"_ in case something still calls next after it was closed. I am learning something new every day...

But it means that this could still be considered, right?

````
  unsubscribe(): void {
    this._value = null;
    super.unsubscribe();
  }
````

**Comment by cartant** [comment]
> But it means that this could still be considered, right?

Yeah. It should probably null it if an error occurs, to - as `getValue` will throw in that scenario, too.

> It also makes me realize that completing subjects for cleaning up is not enough ...

I would not recommend calling `unsubscribe` on subjects. It really has a different meaning in that context; it's really a dispose`.

Rather, I would question why having to call it ought to be necessary in the first place. Did something return a subject to someone else? And said thing wants to guard against `next`, being called? I'd question whether or not that is the preferred way of doing things.

Also, I'd question why it's necessary for the nulling to occur. IMO, it would be indicative of references being held when they ought not to be. Relying upon the nulling of the subject's value to address that would only fix part of the problem.

**Comment by benlesh** [comment]
Hi @Wilt — thank you for taking the time to open this issue and for sharing your thoughtful feedback with the RxJS project. I’m sorry it took us so long to respond. Given how much time has passed and the lack of recent activity, I’m going to close this for now so the tracker reflects the work we’re likely to act on. If this is still relevant on a current RxJS release, please feel welcome to open a new issue with updated context and a reproduction. Thank you again for your contribution and patience.

— Ben
